# Business Problem

# Association Rules

It is a rule-based machine learning technique used to find patterns (relationships, structures) in the data.

Association analysis applications are among the most common applications in data science. It will also coincide as referral systems.

These applications may have come up in the following ways, such as "the person who bought that product also bought this product", "those who viewed that ad also looked at these ads", or "the next video recommended for you".

These scenarios are the most frequently encountered scenarios within the scope of e-commerce data science data mining studies.


# Apriori Algoritması

It is the most used method in this field.

Association rule analysis is carried out by examining some metrics:
X: product
Y: product
N: total trade

*Support: It gives the fraction of transactions which contains product X and Y. Basically Support tells us about the frequently bought products or the combination of products bought frequently.
Support(X, Y) = Freq(X,Y)/N

*Confidence: It tells us how often the products X and Y occur together, given the number times X occurs.
Confidence(X, Y) = Freq(X,Y) / Freq(X)

*Lift: Lift indicates the strength of a rule over the random occurrence of A and B. It basically tells us the strength of any rule.
Lift = Support (X, Y)/( Support(X) * Support(Y) ) 



## 1. Understanding the Data

In [19]:
#!pip install mlxtend

In [20]:
import pandas as pd
import numpy as np

In [21]:
df = pd.read_csv('satislar.csv', names=['products'], header=None)
df

products
0               MILK,BREAD,BISCUIT
1    BREAD,MILK,BISCUIT,CORNFLAKES
2              BREAD,TEA,BOURNVITA
3             JAM,MAGGI,BREAD,MILK
4                MAGGI,TEA,BISCUIT
5              BREAD,TEA,BOURNVITA
6             MAGGI,TEA,CORNFLAKES
7          MAGGI,BREAD,TEA,BISCUIT
8              JAM,MAGGI,BREAD,TEA
9                       BREAD,MILK
10  COFFEE,COCK,BISCUIT,CORNFLAKES
11  COFFEE,COCK,BISCUIT,CORNFLAKES
12          COFFEE,SUGER,BOURNVITA
13               BREAD,COFFEE,COCK
14             BREAD,SUGER,BISCUIT
15         COFFEE,SUGER,CORNFLAKES
16           BREAD,SUGER,BOURNVITA
17              BREAD,COFFEE,SUGER
18              BREAD,COFFEE,SUGER
19      TEA,MILK,COFFEE,CORNFLAKES

In [22]:
df.values

array([['MILK,BREAD,BISCUIT'],
       ['BREAD,MILK,BISCUIT,CORNFLAKES'],
       ['BREAD,TEA,BOURNVITA'],
       ['JAM,MAGGI,BREAD,MILK'],
       ['MAGGI,TEA,BISCUIT'],
       ['BREAD,TEA,BOURNVITA'],
       ['MAGGI,TEA,CORNFLAKES'],
       ['MAGGI,BREAD,TEA,BISCUIT'],
       ['JAM,MAGGI,BREAD,TEA'],
       ['BREAD,MILK'],
       ['COFFEE,COCK,BISCUIT,CORNFLAKES'],
       ['COFFEE,COCK,BISCUIT,CORNFLAKES'],
       ['COFFEE,SUGER,BOURNVITA'],
       ['BREAD,COFFEE,COCK'],
       ['BREAD,SUGER,BISCUIT'],
       ['COFFEE,SUGER,CORNFLAKES'],
       ['BREAD,SUGER,BOURNVITA'],
       ['BREAD,COFFEE,SUGER'],
       ['BREAD,COFFEE,SUGER'],
       ['TEA,MILK,COFFEE,CORNFLAKES']], dtype=object)

In [23]:
df.shape

(20, 1)

## 2. Data Preparation¶

I will use the Apriori algorithm to perform an association analysis.

mlxtend kütüphanesinin apriori metodu, veri setini True-False dataframe olarak kabul etmektedir. Veriyi dönüştürmek için yine mlxtend kütüphanesinin veriyi dönüştürme metodlarını kullanacağım. Bu nedenle elimdeki ham veri setini, bu metodların isteyeceği formata dönüştüreceğim.

In [24]:
# Step1: I converted the data into list format. I separated the objects in each line with ','.

data = list(df["products"].apply(lambda x : x.split(',')))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [25]:
pip install mlxtend

Note: you may need to restart the kernel to use updated packages.


In [26]:
# Step2: I will apply the method of converting the data of mlxtend library into True-False
# dataframe.
# First, I install the mlxtend library for those who do not have it installed.

from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
te_data = te.fit(data).transform(data)
df = pd.DataFrame(te_data,columns=te.columns_)
df

BISCUIT  BOURNVITA  BREAD   COCK  COFFEE  CORNFLAKES    JAM  MAGGI   MILK  \
0      True      False   True  False   False       False  False  False   True   
1      True      False   True  False   False        True  False  False   True   
2     False       True   True  False   False       False  False  False  False   
3     False      False   True  False   False       False   True   True   True   
4      True      False  False  False   False       False  False   True  False   
5     False       True   True  False   False       False  False  False  False   
6     False      False  False  False   False        True  False   True  False   
7      True      False   True  False   False       False  False   True  False   
8     False      False   True  False   False       False   True   True  False   
9     False      False   True  False   False       False  False  False   True   
10     True      False  False   True    True        True  False  False  False   
11     True      False  False   True    True        True  False  False  False   
12    False       True  False  False    True       False  False  False  False   
13    False      False   True   True    True       False  False  False  False   
14     True      False   True  False   False       False  False  False  False   
15    False      False  False  False    True        True  False  False  False   
16    False       True   True  False   False       False  False  False  False   
17    False      False   True  False    True       False  False  False  False   
18    False      False   True  False    True       False  False  False  False   
19    False      False  False  False    True        True  False  False   True   

    SUGER    TEA  
0   False  False  
1   False  False  
2   False   True  
3   False  False  
4   False   True  
5   False   True  
6   False   True  
7   False   True  
8   False   True  
9   False  False  
10  False  False  
11  False  False  
12   True  False  
13  False  False  
14   True  False  
15   True  False  
16   True  False  
17   True  False  
18   True  False  
19  False   True

## 3. Implementing Apriori Algorithm

In the output of the Apriori algorithm, we get the frequencies of each combination in the whole data set. For example, in the output below, only the frequency (frequency) of "BISCUIT" in the whole dataset is 0.35, while the frequency (frequency) of "BISCUIT and BREAD" in the whole dataset is 0.20.

The apriori algorithm was given a min_support value of 0.2. Thus, product associations that are below 0.2 support value in combinations have been eliminated. If the verbose argument is 1, it will tell us how many combinations there are. In our example, 42 combinations were formed. In the last case, we have 16 combinations. Thus, our combination of 42-16 = 26 remained below the value of 0.2 support and was considered as an insignificant rate that we would not add to our comments.

In [27]:
from mlxtend.frequent_patterns import apriori
freq_items = apriori(df,min_support=0.20,use_colnames = True, verbose = 1)
freq_items

Processing 42 combinations | Sampling itemset size 3


support              itemsets
0      0.35             (BISCUIT)
1      0.20           (BOURNVITA)
2      0.65               (BREAD)
3      0.40              (COFFEE)
4      0.30          (CORNFLAKES)
5      0.25               (MAGGI)
6      0.25                (MILK)
7      0.30               (SUGER)
8      0.35                 (TEA)
9      0.20      (BREAD, BISCUIT)
10     0.20         (MILK, BREAD)
11     0.20        (SUGER, BREAD)
12     0.20          (TEA, BREAD)
13     0.20  (COFFEE, CORNFLAKES)
14     0.20       (COFFEE, SUGER)
15     0.20          (MAGGI, TEA)

## 4. Birliktelik Analizi Uygulama

I will apply the association analysis to the combination of mlxtend's association_rules method and the data set that we have support values. I will interpret my latest output according to the values of "support" and "confidence" and suggest a sample action idea.

Interpretation of Sample Association Analysis Output:

The probability of BISCUIT and BREAD being seen together is 20% since support = 0.20.
When BISCUIT is taken, the probability of getting BREAD is around 57% since confidence = 0.571429.

By giving "min_threshold = 0.3", it is ensured that the values with "confidence" value below 0.3 are not brought.

In [28]:
from mlxtend.frequent_patterns import association_rules
df_res = association_rules(freq_items, metric = "confidence", min_threshold = 0.3)
df_res

antecedents   consequents  antecedent support  consequent support  \
0        (BREAD)     (BISCUIT)                0.65                0.35   
1      (BISCUIT)       (BREAD)                0.35                0.65   
2         (MILK)       (BREAD)                0.25                0.65   
3        (BREAD)        (MILK)                0.65                0.25   
4        (SUGER)       (BREAD)                0.30                0.65   
5        (BREAD)       (SUGER)                0.65                0.30   
6          (TEA)       (BREAD)                0.35                0.65   
7        (BREAD)         (TEA)                0.65                0.35   
8       (COFFEE)  (CORNFLAKES)                0.40                0.30   
9   (CORNFLAKES)      (COFFEE)                0.30                0.40   
10      (COFFEE)       (SUGER)                0.40                0.30   
11       (SUGER)      (COFFEE)                0.30                0.40   
12       (MAGGI)         (TEA)                0.25                0.35   
13         (TEA)       (MAGGI)                0.35                0.25   

    support  confidence      lift  leverage  conviction  
0       0.2    0.307692  0.879121   -0.0275    0.938889  
1       0.2    0.571429  0.879121   -0.0275    0.816667  
2       0.2    0.800000  1.230769    0.0375    1.750000  
3       0.2    0.307692  1.230769    0.0375    1.083333  
4       0.2    0.666667  1.025641    0.0050    1.050000  
5       0.2    0.307692  1.025641    0.0050    1.011111  
6       0.2    0.571429  0.879121   -0.0275    0.816667  
7       0.2    0.307692  0.879121   -0.0275    0.938889  
8       0.2    0.500000  1.666667    0.0800    1.400000  
9       0.2    0.666667  1.666667    0.0800    1.800000  
10      0.2    0.500000  1.666667    0.0800    1.400000  
11      0.2    0.666667  1.666667    0.0800    1.800000  
12      0.2    0.800000  2.285714    0.1125    3.250000  
13      0.2    0.571429  2.285714    0.1125    1.750000

## 5. Preparation for Data Filtering

In this section, taking the lowest and highest confidence values, these values will be used in data filtering and the idea of action will be proposed.

Let's find the highest confidence value. The output shows that the highest confidence value is 0.80.

In [32]:
conf_max = df_res['confidence'].max()
conf_max

0.8

Let's find the lowest confidence value. The output shows that the lowest confidence value is 0.307.

In [33]:
conf_min = df_res["confidence"].min()
conf_min

0.3076923076923077

## 6. Data Filtering

Data with the lowest, highest and 0.5 confidence value are filtered. It is ranked in ascending order according to "confidence" value.

In [34]:
df_filt = df_res[ (df_res["confidence"] == conf_min) | (df_res["confidence"] == conf_max) | (df_res["confidence"] == 0.5 )]
df_filt.sort_values("confidence", ascending = True)

antecedents   consequents  antecedent support  consequent support  support  \
0      (BREAD)     (BISCUIT)                0.65                0.35      0.2   
3      (BREAD)        (MILK)                0.65                0.25      0.2   
5      (BREAD)       (SUGER)                0.65                0.30      0.2   
7      (BREAD)         (TEA)                0.65                0.35      0.2   
8     (COFFEE)  (CORNFLAKES)                0.40                0.30      0.2   
10    (COFFEE)       (SUGER)                0.40                0.30      0.2   
2       (MILK)       (BREAD)                0.25                0.65      0.2   
12     (MAGGI)         (TEA)                0.25                0.35      0.2   

    confidence      lift  leverage  conviction  
0     0.307692  0.879121   -0.0275    0.938889  
3     0.307692  1.230769    0.0375    1.083333  
5     0.307692  1.025641    0.0050    1.011111  
7     0.307692  0.879121   -0.0275    0.938889  
8     0.500000  1.666667    0.0800    1.400000  
10    0.500000  1.666667    0.0800    1.400000  
2     0.800000  1.230769    0.0375    1.750000  
12    0.800000  2.285714    0.1125    3.250000

# Comments and Action Advice

Based on the above output;

Line 1:
a). The probability of taking Bread and Biscuits together is 20% as support = 0.2.
b). The probability of getting Biscuits when buying bread is 30% since confidence = 0.307692.
c). The probability of taking the biscuit alone is 35%, since the consequent support = 0.35.

Line 5:
a). The probability of taking coffee and cereal together is 20% as support = 0.2.
b). The probability of taking Cereal when taking coffee is 50% since confidence = 0.5.

Line 7:
a). The probability of taking milk and bread together is 20% since support = 0.2.
b). When milk is taken, the probability of taking Bread is 80% since confidence = 0.8.

Comments :
30% of customers who buy bread also get Biscuits. Cereal was also bought in half of the coffee sales. Milk and Bread were taken together in 20% of all purchases.

Action Proposal:
Customers who buy milk are quite likely to buy bread (80%). This means; customers who buy milk also buy bread, and will likely keep their way to the bread aisle after receiving milk. Milk and bread departments can be located in distant locations within the market, and many departments may be required to pass through the bread department of the customer who purchases milk. Thus, the possibility of purchasing any product during the customer's journey can be increased.

Customers who buy bread seem a bit less likely to buy biscuits (30%). The probability of taking the biscuit alone seems to be low (35%). To increase the sale of biscuits, a biscuit stand can be placed right next to the bread shelf and the ones with cheap prices are placed in the foreground, and the price can be drawn with a dark label to attract the attention of the customer.